src/bin/miri.rs

```diff
@@ -179,6 +196,21 @@ impl rustc_driver::Callbacks for MiriBeRustCompilerCalls {
             });
         }
     }
+
+    fn after_analysis<'tcx>(
+        &mut self,
+        _: &rustc_interface::interface::Compiler,
+        queries: &'tcx rustc_interface::Queries<'tcx>,
+    ) -> Compilation {
+        queries.global_ctxt().unwrap().enter(|tcx| {
+            // Make sure compile_fail tests with post-monomorphization const-eval errors work as
+            // intended in rustdoc mode.
+            if self.mode == MiriBeRustcMode::Rustdoc {
+                let _ = tcx.collect_and_partition_mono_items(());
```

Comment:
So... should we maybe run this always, not just when we are rustdoc? Maybe even when we are Miri and running the program? That makes our builds slower, but it also makes us behave more like real rustc.

Reply:
I can't measure a difference for sysroot builds, even though those are non-trivial crates. So I propose we run `collect_and_partition_mono_items` at least for MIRI_BE_RUSTC mode.

I'm on the fence about whether the interpreter driver itself should do this as well.